### PR TITLE
Addressed Evans comments

### PR DESCRIPTION
**Pull request by zingales**
Except return -1, nil.

### CODE REVIEW

File: routers/routers.go [hunk 1]
```diff
@@ -7,35 +7,39 @@ import (
 	"github.com/atmiguel/cerealnotes/paths"
 )
 
-func SetRoutes() {
+func DefineRoutes() http.Handler {
```
Comment (zingales) [inline routers/routers.go:10]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178422483

File: handlers/handlers.go [hunk 2]
```diff
@@ -171,7 +174,7 @@ func HandleSessionRequest(
 		}
 
 		responseWriter.WriteHeader(http.StatusCreated)
-		responseWriter.Write([]byte(fmt.Sprint("passward email combo was correct")))
+		fmt.Fprint(responseWriter, "passward email combo was correct")
```
Comment (zingales) [inline handlers/handlers.go:177]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178422704

File: handlers/handlers.go [hunk 3]
```diff
@@ -244,16 +248,3 @@ func respondWithMethodNotAllowed(
 	statusCode := http.StatusMethodNotAllowed
 	http.Error(responseWriter, http.StatusText(statusCode), statusCode)
 }
-
```
Comment (zingales) [inline handlers/handlers.go:247]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178423405

File: handlers/handlers.go [hunk 4]
```diff
@@ -131,34 +131,37 @@ func HandleSessionRequest(
...
 	switch request.Method {
 	case http.MethodPost:
-		var loginForm LoginForm
-		body := getRequestBody(request)
+		loginForm := new(LoginForm)
 
-		if err := json.Unmarshal(body, &loginForm); err != nil {
-			panic(err)
+		if err := json.NewDecoder(request.Body).Decode(loginForm); err != nil {
+			http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
```
Comment (zingales) [inline handlers/handlers.go:137]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178423894

File: databaseutil/databaseutil.go [hunk 5]
```diff
@@ -36,24 +36,21 @@ func InsertIntoUsersTable(
 	password []byte,
 	creationTime time.Time,
 ) error {
-	var row *sql.Row
-	{
-		sqlQuery := `
-			INSERT INTO users (display_name, email_address, password, creation_time)
-			VALUES ($1, $2, $3, $4)`
+	sqlQuery := `
```
Comment (zingales) [inline databaseutil/databaseutil.go:39]: Started looking at this code due to: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178424005 and realized we shouldn't be using the query row here since it requires a row so i was able to fix the code by using query instead.

File: handlers/handlers.go [hunk 6]
```diff
@@ -199,7 +202,7 @@ type AuthentictedRequestHandlerType func(
 
 func AuthenticateOrRedirectToLogin(
 	authenticatedHandlerFunc AuthentictedRequestHandlerType,
-) func(http.ResponseWriter, *http.Request) {
+) http.HandlerFunc {
```
Comment (zingales) [inline handlers/handlers.go:205]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178424911

File: handlers/handlers.go [hunk 7]
```diff
@@ -16,8 +14,8 @@ import (
 	"github.com/dgrijalva/jwt-go"
 )
 
-const oneWeekInMinutes = 60 * 24 * 7
-const credentialTimeoutInMinutes = oneWeekInMinutes
+const oneWeek = time.Hour * 24 * 7
```
Comment (zingales) [inline handlers/handlers.go:17]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178424409

File: handlers/tokenutil.go [hunk 8]
```diff
@@ -23,12 +23,12 @@ func parseTokenFromString(tokenAsString string) (*jwt.Token, error) {
...
 func createTokenAsString(
 	userId models.UserId,
-	minutesTilExpiration int64,
+	durationTilExpiration time.Duration,
 ) (string, error) {
 	claims := JwtTokenClaim{
 		userId,
 		jwt.StandardClaims{
-			ExpiresAt: time.Now().Unix() + (minutesTilExpiration * 60),
+			ExpiresAt: time.Now().Add(durationTilExpiration).Unix(),
```
Comment (zingales) [inline handlers/tokenutil.go:31]: https://github.com/atmiguel/cerealnotes/pull/16#discussion_r178424467

**Comment by ethomas2** [comment]
👍

**Review by ethomas2 — APPROVED** [review]

File: databaseutil/databaseutil.go (hunk 5 above)
Comment (atmiguel) [reply databaseutil/databaseutil.go:39]: what do you mean by "it requires a row"?

File: databaseutil/databaseutil.go [hunk 9]
```diff
@@ -36,24 +36,21 @@ func InsertIntoUsersTable(
...
+		INSERT INTO users (display_name, email_address, password, creation_time)
+		VALUES ($1, $2, $3, $4)`
 
-		row = db.QueryRow(sqlQuery, displayName, emailAddress, password, creationTime)
+	rows, err := db.Query(sqlQuery, displayName, emailAddress, password, creationTime)
+	if err != nil {
+		return convertPostgresError(err)
 	}
+	defer rows.Close()
```
Comment (atmiguel) [inline databaseutil/databaseutil.go:47]: why is this necessary? and why wasn't it necessary before?

Comment (atmiguel) [inline databaseutil/databaseutil.go:47]: Shouldn't this be done before the error check?

File: databaseutil/databaseutil.go [hunk 10]
```diff
@@ -36,24 +36,21 @@ func InsertIntoUsersTable(
...
 
-	if err := row.Scan(); err != nil {
-		if err == sql.ErrNoRows { // success
-			return nil
-		}
-
+	if err = rows.Err(); err != nil {
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: random nitpicky thing, but we should do `:=` here. Honestly wish `err` wasn't in scope here, so this couldn't happen, but that doesn't seem worth the hassle.

File: handlers/handlers.go [hunk 11]
```diff
@@ -131,34 +131,37 @@ func HandleSessionRequest(
...
-				panic(err)
+				http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
+				return
 			}
 
-			token, err := createTokenAsString(userId, credentialTimeoutInMinutes)
+			token, err := createTokenAsString(userId, credentialTimeoutDuration)
 			if err != nil {
-				panic(err)
+				http.Error(responseWriter, err.Error(), http.StatusInternalServerError)
```
Comment (atmiguel) [inline handlers/handlers.go:160]: this is the sort of thing we'd ideally handle in some sort of middleware (seeing how often you've had to write it), but this is perfectly good for now

**Review by atmiguel — CHANGES_REQUESTED** [review]

File: databaseutil/databaseutil.go (hunk 5 above)
Comment (zingales) [reply databaseutil/databaseutil.go:39]: i meant it doesn't require a row. QueryRow expects a single row response. An insert does not return rows, and that's why before we had to check for if ErrNoRows, but now i don't need to do that check.

File: databaseutil/databaseutil.go (hunk 9 above)
Comment (zingales) [reply databaseutil/databaseutil.go:47]: I actually did that  originally, but it resulted in an invalid pointer check, because if the first error returns then the rows object is nil. So its only after you know that err == nil that you now rows is defined.

Comment (zingales) [reply databaseutil/databaseutil.go:47]: Honestly it might have been necessary before. I just copied what i saw online when dealing with rows.

File: handlers/handlers.go (hunk 11 above)
Comment (zingales) [reply handlers/handlers.go:160]: Ya i agree, i just didn't know how to write the middleware and and i was like one problem at a time.

File: databaseutil/databaseutil.go (hunk 10 above)
Comment (zingales) [reply databaseutil/databaseutil.go]: I can double check but i think its because err is already defined from the previous statement so i can't := it here.

Comment (atmiguel) [reply databaseutil/databaseutil.go]: I just double-checked. That seems incorrect. Go doesn't have an issue with redefining non-const variables.

**Review by atmiguel — APPROVED** [review]
